jaxopt/_src/linear_solve.py

```diff
@@ -109,14 +109,26 @@ def solve_cg(matvec: Callable,
   return jax.scipy.sparse.linalg.cg(matvec, b, x0=init, **kwargs)[0]
 
 
-def _rmatvec(matvec, x):
-  """Computes A^T x, from matvec(x) = A x, where A is square."""
+def _make_rmatvec(matvec, x):
   transpose = jax.linear_transpose(matvec, x)
-  return transpose(x)[0]
+  return lambda y: transpose(y)[0]
 
 
-def _normal_matvec(matvec, x):
-  """Computes A^T A x from matvec(x) = A x."""
+def _rmatvec(matvec, y, x=None):
+  """Computes rmatvec(y) = A^T y, from matvec(x) = A x.
+
+  If x is None then, we use x = y and assume A is square.
+  """
+  if x is None:
+    x = y
+  rmatvec = _make_rmatvec(matvec, x)
+  return rmatvec(y)
+
+
+# FIXME: potentially rename this utility to _normal_square_matvec.
+# Not doing yet as this name is used in other files from the code base.
+def _normal_matvec(matvec, x=None):
+  """Computes A^T A x from matvec(x) = A x, where A is assumed square."""
   matvec_x, vjp = jax.vjp(matvec, x)
   return vjp(matvec_x)[0]
```

Comment:
Two questions:

1. What is the intended behavior when `x` is `None`?
2. Assuming `x` is not `None`, doesn't this helper work for non-square `A` as well? It reads to me as equivalent to:
```python
  def _normal_matvec2(matvec, x):
    matvec_x = matvec(x)
    return jax.linear_transpose(matvec, x)(matvec_x)
```

Reply:
`x=None` was a leftover of something I had tried before. It was a mistake. Thanks for the catch!